source/common/common/assert.cc

```diff
@@ -0,0 +1,42 @@
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Assert {
+
+class ActionRegistrationImpl : public ActionRegistration {
+public:
+  ActionRegistrationImpl(std::function<void()> action) {
+    ASSERT(debugAssertionFailureRecordAction_ == nullptr);
+    debugAssertionFailureRecordAction_ = action;
+  }
+
+  ~ActionRegistrationImpl() {
+    ASSERT(debugAssertionFailureRecordAction_ != nullptr);
+    debugAssertionFailureRecordAction_ = nullptr;
+  }
+
+  static void invokeAction() {
```

Comment:
Technically if we started other threads before we register the callback, we might have threading issues here. This is somewhat related to my other comment. The way you actually use this, I don't think it matters, but should we add some more comments on current usage and limitations?

Reply:
Adding more comments.